// react-project/lib/webpack.shared.client.config.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.STYLE_LOADER = exports.LOADERS = exports.PLUGINS = exports.BUILD_PATH = exports.ENTRY = undefined;

var _webpack = require('webpack');

var _webpack2 = _interopRequireDefault(_webpack);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _webpackShared = require('./webpack.shared.config');

var SHARED = _interopRequireWildcard(_webpackShared);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ENTRY = exports.ENTRY = {
  app: SHARED.CLIENT_ENTRY,
  _vendor: ['react', 'react-dom', 'react-router', 'react-project']
};

var BUILD_PATH = exports.BUILD_PATH = _path2.default.join(SHARED.APP_PATH, '.build');

var PLUGINS = exports.PLUGINS = SHARED.PLUGINS.concat([new _webpack2.default.optimize.CommonsChunkPlugin('_vendor', 'vendor.js')]);

// ignore all imports from `api` for the client since that code
// only runs on the server
var LOADERS = exports.LOADERS = SHARED.LOADERS.concat([{ test: /modules\/api\//,
  loader: 'null-loader'
}]);

var STYLE_LOADER = exports.STYLE_LOADER = {
  test: SHARED.CSS_REGEX,
  loader: 'style-loader!css-loader?' + SHARED.CSS_LOADER_QUERY + '!postcss-loader',
  name: 'css'
};
